README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> RabbitMQ specific components for Golang

This library is a part of [Pip.Services](https://github.com/pip-services/pip-services) project.
The RabbitMQ module contains a set of components for working with the message queue in RabbitMQ through the AMQP protocol.

The module contains the following packages:
- [**Build**](https://godoc.org/github.com/pip-services3-gox/pip-services3-rabbitmq-gox/build) - Factory for constructing module components
- [**Connect**](https://godoc.org/github.com/pip-services3-gox/pip-services3-rabbitmq-gox/connect) - Components for creating and configuring a connection with RabbitMQ
- [**Queues**](https://godoc.org/github.com/pip-services3-gox/pip-services3-rabbitmq-gox/queues) - Message Queuing components that implement the standard [Messaging](https://github.com/pip-services3-gox/pip-services3-messaging-gox) module interface

<a name="links"></a> Quick links:

* [Configuration](http://docs.pipservices.org/conceptual/configuration/component_configuration/)
* [API Reference](https://godoc.org/github.com/pip-services3-gox/pip-services3-rabbitmq-gox/)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)

## Use

Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-gox/pip-services3-rabbitmq-gox@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.18+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The Golang version of Pip.Services is created and maintained by:
- **Levichev Dmitry**
- **Sergey Seroukhov**

The documentation is written by:
- **Levichev Dmitry**

Many thanks to contibutors, who put their time and talant into making this project better:
- **Andrew Harrington**, Kyrio

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> RabbitMQ specific components for Golang Changelog

## <a name="1.0.0"></a> 1.0.0 (2022-09-21)

### Features
* **build** Factory for constructing module components
* **connect** Components for creating and configuring a connection
* **queues** Message Queuing components

test/queues/MessageQueueFixture.go
package test_queues

import (
	"context"
	"sync"
	"testing"
	"time"

	"github.com/pip-services3-gox/pip-services3-messaging-gox/queues"
	"github.com/stretchr/testify/assert"
)

type MessageQueueFixture struct {
	queue queues.IMessageQueue
}

func NewMessageQueueFixture(queue queues.IMessageQueue) *MessageQueueFixture {
	c := MessageQueueFixture{
		queue: queue,
	}
	return &c
}

func (c *MessageQueueFixture) TestSendReceiveMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	// if c.queue.GetCapabilities().CanMessageCount() {
	// 	count, rdErr := c.queue.MessageCount()
	// 	assert.Nil(t, rdErr)
	// 	assert.Greater(t, count, (int64)(0))
	// }

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestMessageCount(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sendErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sendErr)

	time.Sleep(500 * time.Millisecond)

	count, err := c.queue.ReadMessageCount()
	assert.Nil(t, err)
	assert.True(t, count >= 1)
}

func (c *MessageQueueFixture) TestReceiveSendMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))

	sendErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sendErr)

	time.Sleep(500 * time.Millisecond)

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestReceiveCompleteMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	// count, rdErr := c.queue.ReadMessageCount()
	// assert.Nil(t, rdErr)
	// assert.Greater(t, count, (int64)(0))

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	cplErr := c.queue.Complete(context.Background(), envelope2)
	assert.Nil(t, cplErr)
	assert.Nil(t, envelope2.GetReference())
}

func (c *MessageQueueFixture) TestReceiveAbandonMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	abdErr := c.queue.Abandon(context.Background(), envelope2)
	assert.Nil(t, abdErr)

	envelope2, rcvErr = c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestSendPeekMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	envelope2, pkErr := c.queue.Peek(context.Background(), "")
	assert.Nil(t, pkErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	// pop message from queue for next test
	_, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
}

func (c *MessageQueueFixture) TestPeekNoMessage(t *testing.T) {
	envelope, pkErr := c.queue.Peek(context.Background(), "")
	assert.Nil(t, pkErr)
	assert.Nil(t, envelope)
}

func (c *MessageQueueFixture) TestMoveToDeadMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	mvErr := c.queue.MoveToDeadLetter(context.Background(), envelope2)
	assert.Nil(t, mvErr)
}

func (c *MessageQueueFixture) TestOnMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	receiver := &TestMsgReceiver{}
	c.queue.BeginListen(context.Background(), "", receiver)

	time.Sleep(1000 * time.Millisecond)

	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(1000 * time.Millisecond)

	envelope2 := receiver.GetEnvelope()
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	c.queue.EndListen(context.Background(), "")
}

type TestMsgReceiver struct {
	lock      sync.Mutex
	_envelope *queues.MessageEnvelope
}

func (c *TestMsgReceiver) GetEnvelope() queues.MessageEnvelope {
	c.lock.Lock()
	defer c.lock.Unlock()
	return *c._envelope
}

func (c *TestMsgReceiver) ReceiveMessage(ctx context.Context, envelope *queues.MessageEnvelope, queue queues.IMessageQueue) (err error) {
	c.lock.Lock()
	c._envelope = envelope
	c.lock.Unlock()
	return nil
}


test/queues/RabbitMQMessageQueue_test.go
package test_queues

import (
	"context"
	"os"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	rabbitqueue "github.com/pip-services3-gox/pip-services3-rabbitmq-gox/queues"
	"github.com/stretchr/testify/assert"
)

func TestRabbitMQMessageQueue(t *testing.T) {

	var queue *rabbitqueue.RabbitMQMessageQueue
	var fixture *MessageQueueFixture

	rabbitmqHost := os.Getenv("RABBITMQ_SERVICE_HOST")
	if rabbitmqHost == "" {
		rabbitmqHost = "localhost"
	}
	rabbitmqPort := os.Getenv("RABBITMQ_SERVICE_PORT")
	if rabbitmqPort == "" {
		rabbitmqPort = "5672"
	}

	rabbitmqExchange := os.Getenv("RABBITMQ_EXCHANGE")
	if rabbitmqExchange == "" {
		rabbitmqExchange = "test"
	}

	rabbitmqQueue := os.Getenv("RABBITMQ_QUEUE")
	if rabbitmqQueue == "" {
		rabbitmqQueue = "test"
	}

	rabbitmqUser := os.Getenv("RABBITMQ_USER")
	if rabbitmqUser == "" {
		rabbitmqUser = "user"
	}

	rabbitmqPassword := os.Getenv("RABBITMQ_PASS")
	if rabbitmqPassword == "" {
		rabbitmqPassword = "password"
	}

	if rabbitmqHost == "" && rabbitmqPort == "" {
		return
	}

	queueConfig := cconf.NewConfigParamsFromTuples(
		"exchange", rabbitmqExchange,
		"queue", rabbitmqQueue,
		"options.auto_create", true,
		//"connection.protocol", "amqp",
		"connection.host", rabbitmqHost,
		"connection.port", rabbitmqPort,
		"credential.username", rabbitmqUser,
		"credential.password", rabbitmqPassword,
	)

	queue = rabbitqueue.NewEmptyRabbitMQMessageQueue("testQueue")
	queue.Configure(context.Background(), queueConfig)

	fixture = NewMessageQueueFixture(queue)

	qOpnErr := queue.Open(context.Background(), "")
	if qOpnErr == nil {
		queue.Clear(context.Background(), "")
	}

	defer queue.Close(context.Background(), "")

	t.Run("RabbitMQMessageQueue:Send Receive Message", fixture.TestSendReceiveMessage)
	err := queue.Clear(context.Background(), "")
	assert.Nil(t, err)
	t.Run("RabbitMQMessageQueue:Receive Send Message", fixture.TestReceiveSendMessage)
	err = queue.Clear(context.Background(), "")
	assert.Nil(t, err)
	t.Run("RabbitMQMessageQueue:Receive And Complete Message", fixture.TestReceiveCompleteMessage)
	err = queue.Clear(context.Background(), "")
	assert.Nil(t, err)
	t.Run("RabbitMQMessageQueue:Receive And Abandon Message", fixture.TestReceiveAbandonMessage)
	err = queue.Clear(context.Background(), "")
	assert.Nil(t, err)
	t.Run("RabbitMQMessageQueue:Send Peek Message", fixture.TestSendPeekMessage)
	err = queue.Clear(context.Background(), "")
	assert.Nil(t, err)
	t.Run("RabbitMQMessageQueue:Peek No Message", fixture.TestPeekNoMessage)
	err = queue.Clear(context.Background(), "")
	assert.Nil(t, err)
	t.Run("RabbitMQMessageQueue:On Message", fixture.TestOnMessage)

}


queues/RabbitMQMessageQueue.go
package queues

import (
	"context"
	"time"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cqueues "github.com/pip-services3-gox/pip-services3-messaging-gox/queues"
	mqcon "github.com/pip-services3-gox/pip-services3-rabbitmq-gox/connect"
	rabbitmq "github.com/streadway/amqp"
)

// Message queue that sends and receives messages via MQTT message broker.
//
// MQTT is a popular light-weight protocol to communicate IoT devices.
//
// Configuration parameters:
//
//   - topic:                         name of MQTT topic to subscribe
//
//   connection(s):
//   	- discovery_key:               (optional) a key to retrieve the connection from  IDiscovery
//   	- host:                        host name or IP address
//   	- port:                        port number
//   	- uri:                         resource URI or connection string with all parameters in it
//
//   credential(s):
//   	- store_key:                   (optional) a key to retrieve the credentials from ICredentialStore
//   	- username:                    user name
//   	- password:                    user password
//
// References:
//
// - *:logger:*:*:1.0             (optional) ILogger components to pass log messages
// - *:counters:*:*:1.0           (optional) ICounters components to pass collected measurements
// - *:discovery:*:*:1.0          (optional) IDiscovery services to resolve connections
// - *:credential-store:*:*:1.0   (optional) Credential stores to resolve credentials
//
// Example:
//		ctx := context.Background()
//		queue := queues.NewEmptyRabbitMQMessageQueue("myqueue")
//		queue.Configure(ctx, config.NewConfigParamsFromTuples(
//			"exchange", "my_exchange",
//			"queue", "my_exchange",
//			"options.auto_create", true,
//			"connection.host", "5672",
//			"connection.port", "localhost",
//			"credential.username", "user",
//			"credential.password", "password",
//		))
//		_ = queue.Open(ctx, "123")
//		_ = queue.Send(ctx, "123", cqueues.NewMessageEnvelope("", "mymessage", []byte("ABC")))
//		message, _ := queue.Receive(ctx, "123", 10000*time.Millisecond)
//		if message != nil {
//			// ...
//			queue.Complete(ctx, message)
//		}
//
type RabbitMQMessageQueue struct {
	*cqueues.MessageQueue
	defaultCheckInterval int64
	connection           *rabbitmq.Connection
	mqChanel             *rabbitmq.Channel
	optionsResolver      *mqcon.RabbitMQConnectionResolver
	queue                string
	exchange             string
	exchangeType         string
	routingKey           string
	persistent           bool
	exclusive            bool
	autoCreate           bool
	autoDelete           bool
	noQueue              bool
	Interval             time.Duration

	cancelListen chan bool
	isListen     bool
}

//  Creates a new instance of the message queue.
//	Parameters:
//  	- name(optional) a queue name.
func NewEmptyRabbitMQMessageQueue(name string) *RabbitMQMessageQueue {
	c := RabbitMQMessageQueue{
		defaultCheckInterval: 1000,
		exchange:             "",
		exchangeType:         "fanout",
		routingKey:           "",
		persistent:           false,
		exclusive:            false,
		autoCreate:           false,
		autoDelete:           false,
		noQueue:              false,
		cancelListen:         make(chan bool),
	}

	c.MessageQueue = cqueues.InheritMessageQueue(
		&c, name,
		cqueues.NewMessagingCapabilities(true, true, true, true, true, false, true, false, true))
	c.Interval = time.Duration(c.defaultCheckInterval) * time.Millisecond
	c.optionsResolver = mqcon.NewRabbitMQConnectionResolver()

	return &c
}

func NewRabbitMQMessageQueueFromConfig(name string, config *cconf.ConfigParams) *RabbitMQMessageQueue {
	c := NewEmptyRabbitMQMessageQueue(name)
	if config != nil {
		c.Configure(context.Background(), config)
	}
	return c
}

func NewRabbitMQMessageQueue(name string, mqChanel *rabbitmq.Channel, queue string) *RabbitMQMessageQueue {
	c := NewEmptyRabbitMQMessageQueue(name)
	c.mqChanel = mqChanel
	c.queue = queue
	return c
}

//	Configures component by passing configuration parameters.
//	Parameters:
//		- ctx context.Context
//	- config configuration parameters to be set.
func (c *RabbitMQMessageQueue) Configure(ctx context.Context, config *cconf.ConfigParams) {
	c.MessageQueue.Configure(ctx, config)

	c.Interval = time.Duration(config.GetAsLongWithDefault("interval", int64(c.defaultCheckInterval))) * time.Millisecond

	c.queue = config.GetAsStringWithDefault("queue", c.queue)
	c.exchange = config.GetAsStringWithDefault("exchange", c.exchange)

	c.exchangeType = config.GetAsStringWithDefault("options.exchange_type", c.exchangeType)
	c.routingKey = config.GetAsStringWithDefault("options.routing_key", c.routingKey)
	c.persistent = config.GetAsBooleanWithDefault("options.persistent", c.persistent)
	c.exclusive = config.GetAsBooleanWithDefault("options.exclusive", c.exclusive)
	c.autoCreate = config.GetAsBooleanWithDefault("options.auto_create", c.autoCreate)
	c.autoDelete = config.GetAsBooleanWithDefault("options.auto_delete", c.autoDelete)
	c.noQueue = config.GetAsBooleanWithDefault("options.noqueue", c.noQueue)
}

func (c *RabbitMQMessageQueue) checkOpened(correlationId string) error {
	if c.mqChanel == nil {
		return cerr.NewInvalidStateError(correlationId, "NOT_OPENED", "The queue is not opened")
	}
	return nil
}

//	Checks if the component is opened.
//	Retruns : true if the component has been opened and false otherwise.
func (c *RabbitMQMessageQueue) IsOpen() bool {
	return c.connection != nil && c.mqChanel != nil
}

//	Opens the component with given connection and credential parameters.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
//		- connections connection parameters
//		- credential credential parameters
func (c *RabbitMQMessageQueue) Open(ctx context.Context, correlationId string) error {

	connection, err := c.ConnectionResolver.Resolve(correlationId)
	if err != nil {
		return err
	}

	credential, err := c.CredentialResolver.Lookup(ctx, correlationId)
	if err != nil {
		return err
	}

	options, err := c.optionsResolver.Compose(correlationId, connection, credential)
	if err != nil {
		return err
	}

	if c.queue == "" && c.exchange == "" {
		return cerr.NewConfigError(correlationId,
			"NO_QUEUE",
			"Queue or exchange are not defined in connection parameters")
	}

	conn, err := rabbitmq.Dial(options.GetAsString("uri"))
	if err != nil {
		return err
	}
	c.connection = conn
	c.mqChanel, err = conn.Channel()
	if err != nil {
		return err
	}

	// Automatically create queue, exchange and binding
	if c.autoCreate {
		if c.exchange != "" {
			c.mqChanel.ExchangeDeclare(
				c.exchange,
				c.exchangeType,
				c.persistent,
				c.autoDelete,
				false,
				false,
				nil,
			)
		}

		if !c.noQueue {

			if c.queue == "" {
				res, err := c.mqChanel.QueueDeclare(
					"",
					c.persistent,
					true,
					true,
					false,
					nil,
				)
				if err != nil {
					return err
				}
				c.queue = res.Name
			} else {
				c.mqChanel.QueueDeclare(
					c.queue,
					c.persistent,
					c.exclusive,
					c.autoDelete,
					false,
					nil,
				)
			}

			c.mqChanel.QueueBind(
				c.queue,
				c.routingKey,
				c.exchange,
				false,
				nil,
			)

		}
	}
	return nil
}

// Close mwthod are closes component and frees used resources.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
func (c *RabbitMQMessageQueue) Close(ctx context.Context, correlationId string) (err error) {

	c.Lock.Lock()
	if c.isListen {
		c.isListen = false
		c.Lock.Unlock()
		close(c.cancelListen)
	} else {
		c.Lock.Unlock()
	}

	if c.mqChanel != nil {
		err = c.mqChanel.Close()
		if err != nil {
			return err
		}
	}

	if c.connection != nil {
		err = c.connection.Close()
	}

	c.Lock.Lock()
	defer c.Lock.Unlock()
	c.connection = nil
	c.mqChanel = nil

	c.Logger.Trace(ctx, correlationId, "Closed queue %s", c.queue)
	return err
}

//	ReadMessageCount method are reads the current number of messages in the queue to be delivered.
//	Returns count int64, err error
//	number of messages or error.
func (c *RabbitMQMessageQueue) ReadMessageCount() (count int64, err error) {

	err = c.checkOpened("")
	if err != nil {
		c.Logger.Error(context.Background(), "", err, "RabbitMQMessageQueue:MessageCount: "+err.Error())
		return 0, err
	}

	if c.queue == "" {
		return 0, nil
	}
	queueInfo, err := c.mqChanel.QueueInspect(c.queue)
	if err != nil {
		c.Logger.Error(context.Background(), "", err, "RabbitMQMessageQueue:MessageCount: "+err.Error())
		return 0, err
	}
	return int64(queueInfo.Messages), nil

}

func (c *RabbitMQMessageQueue) toMessage(envelope *rabbitmq.Delivery) *cqueues.MessageEnvelope {
	if envelope == nil {
		return nil
	}

	message := cqueues.MessageEnvelope{
		MessageId:     envelope.MessageId,
		MessageType:   envelope.Type,
		CorrelationId: envelope.CorrelationId,
		Message:       envelope.Body,
		SentTime:      time.Now(),
	}
	message.SetReference(envelope)

	return &message
}

//  Send method are sends a message into the queue.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
//		- message a message envelop to be sent.
func (c *RabbitMQMessageQueue) Send(ctx context.Context, correlationId string, message *cqueues.MessageEnvelope) (err error) {
	err = c.checkOpened(correlationId)
	if err != nil {
		return err
	}

	messageBuffer := rabbitmq.Publishing{
		ContentType: "text/plain",
	}

	if message.CorrelationId != "" {
		messageBuffer.CorrelationId = message.CorrelationId
	}
	if message.MessageId != "" {
		messageBuffer.MessageId = message.MessageId
	}

	if message.MessageType != "" {
		messageBuffer.Type = message.MessageType
	}

	messageBuffer.Body = []byte(message.Message)

	err = c.mqChanel.Publish(c.exchange, c.routingKey, false, false, messageBuffer)

	c.Counters.IncrementOne(ctx, "queue."+c.Name()+".sent_messages")
	c.Logger.Debug(ctx, message.CorrelationId, "Sent message %s via %s", message, c.Name())
	return err
}

//	Peeks a single incoming message from the queue without removing it.
//	If there are no messages available in the queue it returns nil.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
//	Returns: a message
func (c *RabbitMQMessageQueue) Peek(ctx context.Context, correlationId string) (result *cqueues.MessageEnvelope, err error) {
	err = c.checkOpened(correlationId)
	if err != nil {
		return nil, err
	}

	envelope, ok, err := c.mqChanel.Get(c.queue, false)
	if err != nil {
		return nil, err
	}
	if !ok {
		return nil, nil
	}

	message := c.toMessage(&envelope)
	if message != nil {
		c.Logger.Trace(ctx, message.CorrelationId, "Peeked message %s on %s", message, c.Name())
	}

	return message, nil
}

//	PeekBatch method are peeks multiple incoming messages from the queue without removing them.
//	If there are no messages available in the queue it returns an empty list.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
//		- messageCount a maximum number of messages to peek.
//	Returns: a list with messages
func (c *RabbitMQMessageQueue) PeekBatch(ctx context.Context, correlationId string, messageCount int64) (result []*cqueues.MessageEnvelope, err error) {
	err = c.checkOpened(correlationId)
	if err != nil {
		return nil, err
	}
	err = nil
	messages := make([]*cqueues.MessageEnvelope, 0)
	for messageCount > 0 {
		envelope, ok, getErr := c.mqChanel.Get(c.queue, false)
		if getErr != nil || !ok {
			err = getErr
			break
		}
		message := c.toMessage(&envelope)
		messages = append(messages, message)
		messageCount--
	}
	c.Logger.Trace(ctx, correlationId, "Peeked %s messages on %s", len(messages), c.Name())
	return messages, err
}

//  Receive method are receives an incoming message and removes it from the queue.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
//		- waitTimeout a timeout in milliseconds to wait for a message to come.
//	Returns: a message
func (c *RabbitMQMessageQueue) Receive(ctx context.Context, correlationId string, waitTimeout time.Duration) (result *cqueues.MessageEnvelope, err error) {

	err = c.checkOpened(correlationId)
	if err != nil {
		return nil, err
	}
	err = nil

	var message *cqueues.MessageEnvelope
	timeout := waitTimeout

	for {
		if timeout <= 0 {
			break
		}
		// Read the message and exit if received
		env, ok, getErr := c.mqChanel.Get(c.queue, false) // true
		if ok && getErr == nil {
			c.Lock.Lock()
			defer c.Lock.Unlock()
			message = c.toMessage(&env)
			break
		}
		timeout = timeout - c.Interval
	}

	if message != nil {
		c.Counters.IncrementOne(ctx, "queue."+c.Name()+".received_messages")
		c.Logger.Debug(ctx, message.CorrelationId, "Received message %s via %s", message, c.Name())
	}

	return message, nil
}

//	Renews a lock on a message that makes it invisible from other receivers in the queue.
//	This method is usually used to extend the message processing time.
//	Important: This method is not supported by RabbitMQ.
//	Parameters:
//		- ctx context.Context
//		- message a message to extend its lock.
//		- lockTimeout a locking timeout in milliseconds.
func (c *RabbitMQMessageQueue) RenewLock(ctx context.Context, message *cqueues.MessageEnvelope, lockTimeout time.Duration) (err error) {
	// Operation is not supported
	return nil
}

//	Returnes message into the queue and makes it available for all subscribers to receive it again.
//	This method is usually used to return a message which could not be processed at the moment
//	to repeat the attempt.Messages that cause unrecoverable errors shall be removed permanently
//	or/and send to dead letter queue.
//	Important: This method is not supported by RabbitMQ.
//	Parameters:
//		- ctx context.Context
//		- message a message to return.
func (c *RabbitMQMessageQueue) Abandon(ctx context.Context, message *cqueues.MessageEnvelope) (err error) {
	err = c.checkOpened("")
	if err != nil {
		return err
	}
	err = nil

	// Make the message immediately visible
	envelope, ok := message.GetReference().(*rabbitmq.Delivery)
	if ok {
		err = c.mqChanel.Nack(envelope.DeliveryTag, false, true)
		if err != nil {
			return err
		}
		message.SetReference(nil)
		c.Logger.Trace(ctx, message.CorrelationId, "Abandoned message %s at %c", message, c.Name())
	}
	return nil
}

//	Permanently removes a message from the queue.
//	This method is usually used to remove the message after successful processing.
//	Important: This method is not supported by RabbitMQ.
//	Parameters:
//		- ctx context.Context
//		- message a message to remove.
func (c *RabbitMQMessageQueue) Complete(ctx context.Context, message *cqueues.MessageEnvelope) (err error) {
	err = c.checkOpened("")
	if err != nil {
		return err
	}
	err = nil
	envelope, ok := message.GetReference().(*rabbitmq.Delivery)
	if ok {
		c.mqChanel.Ack(envelope.DeliveryTag, false)
		message.SetReference(nil)
		c.Logger.Trace(ctx, message.CorrelationId, "Completed message %s at %s", message, c.Name())
	}
	return nil
}

//	Permanently removes a message from the queue and sends it to dead letter queue.
//	Important: This method is not supported by RabbitMQ.
//	Parameters:
//		- ctx context.Context
//		- message a message to be removed.
//	Returns: error
func (c *RabbitMQMessageQueue) MoveToDeadLetter(ctx context.Context, message *cqueues.MessageEnvelope) (err error) {
	// Operation is not supported
	return nil
}

//	Listens for incoming messages and blocks the current thread until queue is closed.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
//		- callback
//  Returns: listen error
func (c *RabbitMQMessageQueue) Listen(ctx context.Context, correlationId string, receiver cqueues.IMessageReceiver) error {
	err := c.checkOpened("")
	if err != nil {
		c.Logger.Error(ctx, correlationId, err, "RabbitMQMessageQueue:Listen: Can't start listen "+err.Error())
		return nil
	}

	c.Logger.Debug(ctx, correlationId, "Started listening messages at %s", c.Name())

	messageChannel, err := c.mqChanel.Consume(
		c.queue,
		c.exchange,
		false,
		false,
		false,
		false,
		nil,
	)

	if err != nil {
		c.Logger.Error(ctx, correlationId, err, "RabbitMQMessageQueue:Listen: Can't consume to queue"+err.Error())
		return nil
	}

	c.EndListen(ctx, correlationId)
	c.cancelListen = make(chan bool)

	go func() {
		c.Lock.Lock()
		c.isListen = true
		c.Lock.Unlock()

		for {
			select {
			case msg := <-messageChannel:
				{
					message := c.toMessage(&msg)
					c.Counters.IncrementOne(ctx, "queue."+c.Name()+".received_messages")
					c.Logger.Debug(ctx, message.CorrelationId, "Received message %s via %s", message, c.Name())
					recvErr := receiver.ReceiveMessage(ctx, message, c)
					if recvErr != nil {
						c.Logger.Error(ctx, message.CorrelationId, recvErr, "Processing received message %s error in queue %s", message, c.Name())
					}
					c.mqChanel.Ack(msg.DeliveryTag, false)
				}
			case <-c.cancelListen:
				{
					return
				}
			case <-ctx.Done():
				{
					return
				}
			}
		}
	}()

	return nil
}

//	Ends listening for incoming messages.
//	When this method is call listen unblocks the thread and execution continues.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
func (c *RabbitMQMessageQueue) EndListen(ctx context.Context, correlationId string) {
	c.Lock.Lock()
	if c.isListen {
		c.isListen = false
		c.Lock.Unlock()
		close(c.cancelListen)
	} else {
		c.Lock.Unlock()
	}
}

//	Clear method are clears component state.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
//  Returns: error
func (c *RabbitMQMessageQueue) Clear(ctx context.Context, correlationId string) (err error) {
	err = c.checkOpened("")
	if err != nil {
		return err
	}

	count := 0
	if c.queue != "" {
		count, err = c.mqChanel.QueuePurge(c.queue, false)
	}

	if err != nil {
		return err
	}

	c.Logger.Trace(ctx, correlationId, "Cleared  %s messages in queue %s", count, c.Name())

	return nil
}
